### PR TITLE
feat:support ansi mode rounding function

### PR DESCRIPTION
**Pull request by coderfender**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion-comet/issues/466

## Rationale for this change

Support ANSI mode as part of  https://github.com/apache/datafusion-comet/issues/313 (link to EPIC) . 

## What changes are included in this PR?

1. Changes to proto definition for scalar function to add a new parameter called `fail_on_error` in coherence with Spark definitions. 
2. Changes to `CometRound` to pass `fail_on_error` respecting ANSI mode 
3. Update other scalar functions with `false` input to ensure existing functionality is unaffected
4. Changes in `planner.rs` to respect `fail_on_error` param
5. Changes in `spark_round` function (specifically the `round_integer` macros) to respect ANSI mode (`fail_on_error`) and throw overflow error in case rounding op fails
6. Unit tests on the scala side to test edge cases (with / without ANSI mode enabled) 

## How are these changes tested?

Unit tests on the scala side to test edge cases (with / without ANSI mode enabled)

### CODE REVIEW

**Comment by coderfender** [comment]
@andygrove  , This is the PR to support ANSI mode with rounding functions. Please take a look whenever you get a chance.

File: native/spark-expr/src/math_funcs/round.rs [hunk 1]
```diff
@@ -15,49 +15,90 @@
...
         let rem = $X % $DIV;
         if rem <= -$HALF {
-            ($X - rem).sub_wrapping($DIV)
+            if $FAIL_ON_ERROR {
+                match ($X - rem).sub_checked($DIV) {
+                    Ok(v) => Ok(v),
+                    Err(_e) => Err(ArrowError::ComputeError(
+                        arithmetic_overflow_error("integer").to_string(),
+                    )),
+                }
```
Comment (andygrove) [inline native/spark-expr/src/math_funcs/round.rs]: Since the `Ok` value is unchanged and the `Err` is changed, it would be better to use `map_err` rather than a match here.

```suggestion
                ($X - rem).sub_checked($DIV).map_err(|_| {
                    ArrowError::ComputeError(arithmetic_overflow_error("integer").to_string())
                })
```

File: native/spark-expr/src/math_funcs/round.rs [hunk 2]
```diff
@@ -15,49 +15,90 @@
...
+            } else {
+                Ok(($X - rem).sub_wrapping($DIV))
+            }
         } else if rem >= $HALF {
-            ($X - rem).add_wrapping($DIV)
+            if $FAIL_ON_ERROR {
+                match ($X - rem).add_checked($DIV) {
```
Comment (andygrove) [inline native/spark-expr/src/math_funcs/round.rs]: `map_err` should be used here as well

File: native/spark-expr/src/math_funcs/round.rs [hunk 3]
```diff
@@ -15,49 +15,90 @@
...
+                        arithmetic_overflow_error("integer").to_string(),
+                    )),
+                }
+            } else {
+                Ok(($X - rem).add_wrapping($DIV))
+            }
         } else {
-            $X - rem
+            if $FAIL_ON_ERROR {
+                match $X.sub_checked(rem) {
```
Comment (andygrove) [inline native/spark-expr/src/math_funcs/round.rs]: `map_err` should be used here as well

File: native/spark-expr/src/math_funcs/round.rs (hunk 1 above)
Comment (coderfender) [reply native/spark-expr/src/math_funcs/round.rs]: Thank you for the suggestion @andygrove  . Let me update it to use `map_err` instead of `match` statement

File: native/spark-expr/src/math_funcs/round.rs (hunk 2 above)
Comment (coderfender) [reply native/spark-expr/src/math_funcs/round.rs]: Sure

File: native/spark-expr/src/math_funcs/round.rs (hunk 3 above)
Comment (coderfender) [reply native/spark-expr/src/math_funcs/round.rs]: Sure

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 4]
```diff
@@ -3017,6 +3017,37 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
+  test("ANSI support for round function") {
+    val data = Seq((Integer.MAX_VALUE, Integer.MIN_VALUE, Long.MinValue, Long.MaxValue))
+    Seq("true", "false").foreach { p =>
+      withSQLConf(SQLConf.ANSI_ENABLED.key -> p) {
+        withParquetTable(data, "tbl") {
+          val res = spark.sql(s"""
+                                |SELECT
+                                |  round(_1, -1) ,
+                                |  round(_1, -10) ,
+                                |  round(${Int.MaxValue}, -10)
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: could you also add tests where scale is 0 and >0.

Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: The test should really just include one instance of `round` per query, otherwise we don't really know which one is throwing an exception.

Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: It would also be better to test the input values individually. I'd suggest something like this:

```scala
  test("ANSI support for round function") {
    for (value <- Seq((Integer.MAX_VALUE, Integer.MIN_VALUE, Long.MinValue, Long.MaxValue))) {
      val data = Seq(value)
      withParquetTable(data, "tbl") {
        for (scale <- Seq(-10, -1, 0, 1, 10)) {
          for (ansi <- Seq(true, false)) {
            withSQLConf(SQLConf.ANSI_ENABLED.key -> ansi.toString) {
              val res = spark.sql(s"SELECT round(_1, $scale) from tbl")
              checkSparkMaybeThrows(res) match {
                case (Some(sparkException), Some(cometException)) =>
                  assert(sparkException.getMessage.contains("ARITHMETIC_OVERFLOW"))
                  assert(cometException.getMessage.contains("ARITHMETIC_OVERFLOW"))
                case (None, None) => checkSparkAnswerAndOperator(res)
                case (None, Some(ex)) =>
                  fail(
                    "Comet threw an exception but Spark did not. Comet exception: " + ex.getMessage)
                case (Some(sparkException), None) =>
                  fail(
                    "Spark threw an exception but Comet did not. Spark exception: " +
                      sparkException.getMessage)
              }
            }
          }
        }
      }
    }
  }
```

Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: Done

**Comment by coderfender** [comment]
@andygrove , I made the changes per review . Please take a look whenever you get a chance

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 5]
```diff
@@ -3017,6 +3017,32 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("ANSI support for round function") {
+    Seq((Integer.MAX_VALUE, Integer.MIN_VALUE, Long.MinValue, Long.MaxValue)).foreach { value =>
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: Could you also add Byte and Short min and max

Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: Sure @andygrove

Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: For consistency, it would be good to stick to Scala's constants for min and max, rather than mixing Java and Scala versions.

```suggestion
    Seq((Int.MaxValue Int.MinValue Long.MinValue, Long.MaxValue)).foreach { value =>
```

**Comment by andygrove** [comment]
@coderfender Could you also updatre the documentation (list of supported expressions)

**Comment by coderfender** [comment]
Sure let me update the documentation and the test cases @andygrove

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala (hunk 5 above)
Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: Sure @andygrove

**Comment by coderfender** [comment]
Thank you for the approval @andygrove
